### ISSUE TITLE
Incorrect error position on resolving overload with a nullable argument on non-nullable parameters

### ISSUE THREAD

**Issue by tinganho** [Microsoft/TypeScript#13198]
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.201xxxxx)

**Code**
On `strictNullCheck`:
```ts
function replace(p1: string, p2: string): void;
function replace(p1: RegExp, p2: string): void;
function replace(p1: RegExp | string, p2: string) { }

let p: string | null;
function f() {
  replace('pd', p); // Error on '"pd"', instead of 'p'.
}
```

**Expected behavior:**
Error on `p`. I think the compiler should be smart enough to give feedback on `p` is nullable and not mark `"pd"` as an error. The error is quite confusing since `"pd"` is correct.
**Actual behavior:**
Error on `"pd"`.

**Comment by mhegazy** [comment]
This is a result of the overload resolution/error reporting process, and not specific to strictNullchecks. the overloads are tried in order, the error is reported on the last one, as it is assumed to be more general, not in this case.

The last overload has first parameter as `RegExp`, and `string` is not a `RegExp`.

I am guessing you meant to add a third overload `replace(p1: RegExp | string, p2: string)`? if so, then the error would be issued on the second parameter as you would expect.

as a side note, this function really needs one signature, and not 2:

```ts
function replace(p1: RegExp | string, p2: string): void { }
```

**Comment by tinganho** [comment]
@mhegazy I guess it was the current overload resolution.

Maybe it is too much to ask, but couldn't it, on no matching overload, error on the one that "fits" the most? Like in the above example, a comparison in the second overload, there is a mismatch in the first parameter between the types `string` and `RegExp`, which is a "stronger" mismatch than the first overload, which has a mismatch between the types in the second parameter `string` and `string | null`. The assigned type match one of the constituents in `string | null`, so it is a weaker mismatch than the second overload.

Because the first overload is a _weaker mismatch_ and probably the one that fits the most, it should use that one to error on.

_(first overload = top overload)_

**Comment by mhegazy** [comment]
> Maybe it is too much to ask, but couldn't it, on no matching overload, error on the one that "fits" the most? 

We are open to suggestions for an algorithm that can make the error reporting better.
